File: php/easypay.lib/Easypay_request.php
<?php

namespace easpaylib;

class Easypay_request
{
    const newLine = "\n";

    /**
     * create data-string for signature
     *
     * @param $http_request_method string
     * @param $content_hash       string
     * @param $content_type       string
     * @param $date              string
     * @param $path              string
     * @return string
     */
    public static function create_hash_string($http_request_method, $content_hash, $content_type, $date, $path)
    {
        return $http_request_method . self::newLine . $content_hash . self::newLine . $content_type . self::newLine . $date . self::newLine . $path;
    }

    /**
     * @param $config Easypay_config
     * @param $payment_request Checkout_page_request
     * @return string
     */
    public static function get_checkout_page_url($config, $payment_request)
    {
        $data = json_encode($payment_request->get_json_data());
        $signature = Signature::sign($data, $config->get_easypay_secret());
        $path = "/authorize.jsf?signature=" . rawurlencode(base64_encode($signature)) . "&checkoutRequestItem=" . rawurlencode(base64_encode($data));
        return "http://" . $config->get_host() . $config->get_basepath() . $path;
    }

    /**
     * @param $payment_id string
     * @param $easypay_config Easypay_config
     * @param $req_id string
     * @return Payment_response
     */
    public static function commit_payment_request($payment_id, $easypay_config, $req_id)
    {
        return self::_process_request($easypay_config, json_encode(array("operation" => "COMMIT")), "PUT", "/payments/" . $payment_id, $req_id);
    }

    /**
     * @param $payment_id string
     * @param $easypay_config Easypay_config
     * @param $req_id string
     * @return Payment_response
     */
    public static function cancel_payment_request($payment_id, $easypay_config, $req_id)
    {
        return self::_process_request($easypay_config, json_encode(array("operation" => "CANCEL")), "PUT", "/payments/" . $payment_id, $req_id);
    }

    /**
     * @param $payment_id string
     * @param $easypay_config Easypay_config
     * @param $req_id string
     * @return Status_response
     */
    public static function get_payment_status($payment_id, $easypay_config, $req_id)
    {
        return self::_process_request($easypay_config, null, "GET", "/payments/" . $payment_id, $req_id);
    }

    /**
     * @param $config Easypay_config
     * @param $data string
     * @param $http_request_method string
     * @param $path string
     * @param $req_id string
     * @return mixed
     */
    private static function _process_request($config, $data, $http_request_method, $path, $req_id)
    {
        $content_type = ($http_request_method === 'GET') ? '' : 'application/vnd.ch.swisscom.easypay.direct.payment+json';
        $accept = ($http_request_method === 'GET') ? "application/vnd.ch.swisscom.easypay.direct.payment+json" : "application/vnd.ch.swisscom.easypay.message.list+json";
        $date = gmdate('D, d M Y H:i:s +0000', time());
        $url = "https://" . $config->get_host() . $config->get_basepath() . $path;

        $md5Hash = isset($data) ? base64_encode(Signature::hash($data)) : "";
        $hash_string = self::create_hash_string($http_request_method, $md5Hash, $content_type, $date, $path);
        $signature = Signature::sign($hash_string, $config->get_easypay_secret()); //Signature::Sign($hashString, base64_decode($config->getEasypaySecret()));

        $headers = array(
            "Content-Type: " . $content_type,
            "X-SCS-Date: " . $date,
            "X-Request-Id: " . $req_id,
            "X-Merchant-Id: " . $config->get_merchant_id(),
            "X-CE-Client-Specification-Version: 1.1",
            "X-SCS-Signature: " . base64_encode($signature),
            "Accept: " . $accept,
            "Date: " . $date,
        );

        if (isset($data))
            array_push($headers, "Content-MD5: " . $md5Hash);

        $options = array(
            CURLOPT_CUSTOMREQUEST => $http_request_method,
            CURLOPT_RETURNTRANSFER => 1,
            CURLOPT_HEADER => 0,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_POSTFIELDS => $data,
            CURLOPT_SSL_VERIFYHOST => 0,
            CURLOPT_SSL_VERIFYPEER => 0
        );

        $ch = curl_init($url);
        curl_setopt_array($ch, $options);

        $response = curl_exec($ch);
        $http_code = curl_getinfo($ch)["http_code"];

        $payment_response = new Payment_response();

        echo "response: " . $response . "<br/>";

        if (curl_errno($ch) != 0 || $http_code !== 200) {
            $payment_response->setSuccess(false);

            if (curl_errno($ch) != 0) {
                $error = curl_error($ch);
                $payment_response->set_exception($error);
            } else {
                $messages = json_decode($response, true);

                foreach ($messages["messages"] as $msg) {
                    $message_obj = new Message();
                    if (array_key_exists("code", $msg))
                        $message_obj->set_code($msg["code"]);
                    if (array_key_exists("field", $msg))
                        $message_obj->set_field($msg["field"]);
                    if (array_key_exists("message", $msg))
                        $message_obj->set_message($msg["message"]);
                    if (array_key_exists("requestId", $msg))
                        $message_obj->set_request_id($msg["requestId"]);

                    $payment_response->add_message($message_obj);
                }
            }

        } else {
            $payment_response->setSuccess(true);
            $statusResponse = json_decode($response, true);

            if (isset($statusResponse)) {

                $payment_response = new Status_response();
                if (array_key_exists("orderID", $statusResponse))
                    $payment_response->set_orderID($statusResponse["orderID"]);
                if (array_key_exists("paymentInfo", $statusResponse))
                    $payment_response->set_payment_info($statusResponse["paymentInfo"]);
                if (array_key_exists("isSilentAuthenticated", $statusResponse))
                    $payment_response->set_is_silent_authenticated($statusResponse["isSilentAuthenticated"]);
                if (array_key_exists("amount", $statusResponse))
                    $payment_response->set_amount($statusResponse["amount"]);
                if (array_key_exists("msisdn", $statusResponse))
                    $payment_response->set_msisdn($statusResponse["msisdn"]);
                if (array_key_exists("status", $statusResponse))
                    $payment_response->set_status($statusResponse["status"]);
                if (array_key_exists("formattedMsisdn", $statusResponse))
                    $payment_response->set_formatted_msisdn($statusResponse["formattedMsisdn"]);
            }
        }

        curl_close($ch);

        return $payment_response;
    }
}

File: Android/EasypayClient/clientlibrary/src/main/java/com/swisscom/easypay/clientlibrary/EasypayRequest.java
package com.swisscom.easypay.clientlibrary;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLEncoder;
import java.security.SignatureException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

import android.util.Base64;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class EasypayRequest {
    /**
     * create data-string for signature
     *
     * @param httpRequestMethod String
     * @param contentHash       String
     * @param contentType       String
     * @param date              String
     * @param path              String
     * @return String
     */
    private static String createHashString(String httpRequestMethod, String contentHash, String contentType, String date, String path) {
        return httpRequestMethod + '\n' + contentHash + '\n' + contentType + '\n' + date + '\n' + path;
    }

    public static URL GetCheckoutPageUrl(EasypayConfig config, CheckoutPageRequest paymentRequest) {
        try {
            String data = paymentRequest.toJSONObject().toString();
            byte[] dataByteArray = data.getBytes(Signature.DEFAULT_ENCODING);
            byte[] key = config.getEasypaySecret().getBytes(Signature.DEFAULT_ENCODING);
            byte[] signature = Signature.sign(dataByteArray, key);
            String path = "/authorize.jsf?signature=" + URLEncoder.encode(Base64.encodeToString(signature, Base64.NO_WRAP), Signature.DEFAULT_ENCODING) + "&checkoutRequestItem=" + URLEncoder.encode(Base64.encodeToString(dataByteArray, Base64.NO_WRAP), Signature.DEFAULT_ENCODING);
            return new URL("http://" + config.getHost() + config.getBasepath() + path);
        } catch (MalformedURLException | SignatureException | UnsupportedEncodingException | JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static PaymentResponse CommitPaymentRequest(String paymentId, EasypayConfig easypayConfig, String reqId) {
        String dataJSON = "{\"operation\":\"COMMIT\"}";
        String path = "/payments/" + paymentId;
        return ProcessPaymentRequest(dataJSON, path, easypayConfig, reqId);
    }

    public static PaymentResponse CancelPaymentRequest(String paymentId, EasypayConfig easypayConfig, String reqId) {
        String dataJSON = "{\"operation\":\"CANCEL\"}";
        String path = "/payments/" + paymentId;

        return ProcessPaymentRequest(dataJSON, path, easypayConfig, reqId);
    }

    public static StatusResponse GetPaymentStatus(String paymentId, EasypayConfig easypayConfig, String reqId) {
        Response result = ProcessRequest(easypayConfig, null, "GET", "/payments/" + paymentId, reqId);
        System.out.println("Response: " + result.getResult());

        StatusResponse statusResponse = new StatusResponse();
        try {
            JSONObject jObject = new JSONObject(result.getResult());

            if (jObject.has("amount"))
                statusResponse.setAmount(jObject.getString("amount"));
            if (jObject.has("formattedMsisdn"))
                statusResponse.setFormattedMsisdn(jObject.getString("formattedMsisdn"));
            if (jObject.has("isSilentAuthenticated"))
                statusResponse.setIsSilentAuthenticated(jObject.getBoolean("isSilentAuthenticated"));
            if (jObject.has("msisdn"))
                statusResponse.setMsisdn(jObject.getString("msisdn"));
            if (jObject.has("orderID"))
                statusResponse.setOrderID(jObject.getString("orderID"));
            if (jObject.has("paymentInfo"))
                statusResponse.setPaymentInfo(jObject.getString("paymentInfo"));
            if (jObject.has("status"))
                statusResponse.setStatus(jObject.getString("status"));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return statusResponse;
    }

    public static PaymentResponse ProcessPaymentRequest(String dataJson, String path, EasypayConfig easypayConfig, String reqId) {
        Response result = ProcessRequest(easypayConfig, dataJson, "PUT", path, reqId);
        PaymentResponse paymentResponse = new PaymentResponse();
        System.out.println("Response: " + result.getResult());

        paymentResponse.setSuccess(result.isSuccess());

        if (!result.isSuccess() && result.getException() == null) {
            JSONObject jObject = null;
            try {
                jObject = new JSONObject(result.getResult());

                JSONArray messageArray = jObject.getJSONArray("messages");

                for (int i = 0; i < messageArray.length(); i++) {

                    Message msg = new Message();
                    JSONObject messageObject = messageArray.getJSONObject(i);
                    if (messageObject.has("code"))
                        msg.setCode(messageObject.getString("code"));
                    if (messageObject.has("field"))
                        msg.setField(messageObject.getString("field"));
                    if (messageObject.has("message"))
                        msg.setMessage(messageObject.getString("message"));
                    if (messageObject.has("requestId"))
                        msg.setRequestId(messageObject.getString("requestId"));

                    paymentResponse.getMessages().add(msg);

                }

            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        return paymentResponse;
    }

    private static Response ProcessRequest(EasypayConfig config, String data, String httpRequestMethod, String path, String reqId) {
        String contentType = httpRequestMethod.equals("GET") ? "" : "application/vnd.ch.swisscom.easypay.direct.payment+json";

        String dateFormat = "EEE, dd MMM yyyy HH:mm:ss Z";
        SimpleDateFormat df = new SimpleDateFormat(dateFormat, Locale.ENGLISH);
        String date = df.format(Calendar.getInstance().getTime());

        System.out.println("date: " + date);

        Response response = new Response();
        HttpURLConnection urlConnection = null;
        try {
            URL url = new URL("https://" + config.getHost() + config.getBasepath() + path);

            urlConnection = (HttpURLConnection) url.openConnection();

            urlConnection.setRequestMethod(httpRequestMethod);
            urlConnection.setRequestProperty("X-SCS-Date", date);
            urlConnection.setRequestProperty("X-Request-Id", reqId);
            urlConnection.setRequestProperty("X-Merchant-Id", config.getMerchantId());
            urlConnection.setRequestProperty("X-CE-Client-Specification-Version", "1.1");

            urlConnection.setRequestProperty("Accept", "application/vnd.ch.swisscom.easypay.message.list+json");
            urlConnection.setRequestProperty("Date", date);

            boolean hasData = data != null;

            byte[] dataByteArray = hasData ? data.getBytes(Signature.DEFAULT_ENCODING) : null;
            byte[] key = config.getEasypaySecret().getBytes(Signature.DEFAULT_ENCODING);
            byte[] md5Hash = hasData ? Signature.hash(dataByteArray) : new byte[0];

            String hashString = createHashString(httpRequestMethod, Base64.encodeToString(md5Hash, Base64.NO_WRAP), contentType, date, path);
            byte[] signature = Signature.sign(hashString.getBytes(Signature.DEFAULT_ENCODING), key);

            urlConnection.setRequestProperty("X-SCS-Signature", Base64.encodeToString(signature, Base64.NO_WRAP));

            urlConnection.setUseCaches(false);
            urlConnection.setDoInput(true);

            if (hasData) {
                urlConnection.setRequestProperty("Content-Type", contentType);
                urlConnection.setRequestProperty("Content-MD5", Base64.encodeToString(md5Hash, Base64.NO_WRAP));

                urlConnection.setDoOutput(true);

                OutputStream out = new BufferedOutputStream(urlConnection.getOutputStream());
                out.write(dataByteArray);
                out.close();
            }

            urlConnection.connect();

            StringBuilder sb = new StringBuilder();

            int HttpResult = urlConnection.getResponseCode();

            response.setSuccess(HttpResult == HttpURLConnection.HTTP_OK);
            BufferedReader br;

            if (HttpResult != HttpURLConnection.HTTP_OK) {
                br = new BufferedReader(new InputStreamReader(urlConnection.getErrorStream(), "utf-8"));
            } else {
                br = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "utf-8"));
            }
            String line;
            while ((line = br.readLine()) != null) {
                sb.append(line).append("\n");
            }
            br.close();

            response.setResult(sb.toString());

        } catch (Exception e) {
            response.setException(e);
            e.printStackTrace();
        } finally {
            if (urlConnection != null)
                urlConnection.disconnect();
        }

        return response;
    }
}

class Response {
    private boolean success;
    private Exception exception;
    private String result;

    public boolean isSuccess() {
        return success;
    }

    public void setSuccess(boolean success) {
        this.success = success;
    }

    public void setException(Exception exception) {
        this.success = false;
        this.exception = exception;
    }

    public Exception getException() {
        return exception;
    }

    public void setResult(String result) {
        this.result = result;
    }

    public String getResult() {
        return result;
    }
}

